android/src/main/java/com/jhomlala/better_player/BetterPlayer.java

```diff
@@ -122,6 +125,71 @@
         setupVideoPlayer(eventChannel, textureEntry, result);
     }
 
+    static void preCache(Context context, String key, String dataSource, long preCacheSize, long maxCacheSize, long maxCacheFileSize, Map<String, String> headers, Result result) {
+
+        Uri uri = Uri.parse(dataSource);
+        DataSource.Factory dataSourceFactory;
+
+        String userAgent = System.getProperty(USER_AGENT_PROPERTY);
+        if (headers != null && headers.containsKey(USER_AGENT)) {
+            String userAgentHeader = headers.get(USER_AGENT);
+            if (userAgentHeader != null) {
+                userAgent = userAgentHeader;
+            }
+        }
+
+        if (isHTTP(uri)) {
+            dataSourceFactory = new DefaultHttpDataSource.Factory()
+                    .setUserAgent(userAgent)
+                    .setAllowCrossProtocolRedirects(true)
+                    .setConnectTimeoutMs(DefaultHttpDataSource.DEFAULT_CONNECT_TIMEOUT_MILLIS)
+                    .setReadTimeoutMs(DefaultHttpDataSource.DEFAULT_READ_TIMEOUT_MILLIS);
+
+            if (headers != null) {
+                ((DefaultHttpDataSource.Factory) dataSourceFactory).setDefaultRequestProperties(headers);
+            }
+
+
+            DataSpec dataSpec = new DataSpec(uri, 0, preCacheSize);
+
+            class ProgressListener implements CacheWriter.ProgressListener {
+                @Override
+                public void onProgress(long requestLength, long bytesCached, long newBytesCached) {
+                    if (bytesCached >= requestLength) {
+                        //notify the requested size has finished caching
+                        new Handler(Looper.getMainLooper()).post(() -> result.success(null));
+                    }
+                }
+            }
+            CacheDataSourceFactory cacheDataSourceFactory =
+                    new CacheDataSourceFactory(context, maxCacheSize, maxCacheFileSize, dataSourceFactory);
+
+
+            CacheWriter cacheWriter = new CacheWriter(cacheDataSourceFactory.createDataSource(), dataSpec, true, null, new ProgressListener());
+
+            //start caching in a background thread
+            Runnable r = () -> {
+                try {
+                    cacheWriter.cache();
+                } catch (Exception e) {
+                    //we have to catch HttpDataSourceException manually to avoid throwing when the video is actually fully loaded
+                    //see https://github.com/google/ExoPlayer/issues/7326
+                    if (e instanceof HttpDataSourceException) {
+                        new Handler(Looper.getMainLooper()).post(() -> result.success(null));
+                    } else {
+                        //notify caching failed
+                        new Handler(Looper.getMainLooper()).post(() -> result.error(e.toString(), Arrays.toString(e.getStackTrace()), e));
+                    }
+                }
+            };
+            new Thread(r).start();
```

Comment:
You're starting thread, but later you're not managing it. There's no try/catch and you're not tear down this trhead.

Reply:
What do you mean managing it? There is a timeout on the request and once the thread is completed (or failed), it will die alone. No need to call join() or anything. Error handling is done inside of the thread already. If you want to change it, feel free to submit a change request. It LGTM as is.